### PR TITLE
Variations: Add error state for syncing global attribute options

### LINKED ISSUE #3601: Variations: Implement error notice while syncing global attribute options
There is the possibility that the sync operation fails. We need to inform the user and give a retry option

Tasks
- [ ] If the sync fails show notice with a retry option.

### PR DESCRIPTION
**Pull request by ealeksandrov**
Closes https://github.com/woocommerce/woocommerce-ios/issues/3601.

## Description

This PR adds error state on attribute options screen.

## Changes

1. Switch `isSyncing` from `Bool` to `SyncState` enum.
2. Add `.failed` case for `SyncState`
3. Display `EmptyStateViewController` with error description and retry action for `.failed` case in `AddAttributeOptionsViewController`.

`EmptyStateViewController` picked instead of non-blocking `Notice` because we need to block user from adding new options until existing data is properly synced.

## Test

1. Open variable product, with existing variations or not
2. Tap "variations" (or "add variation")
3. For no variations:
a. Tap "Add Variation"
For existing variations:
a. Tap "•••" in navbar, pick "Edit Attributes"
b. Tap "Add New Attribute"

4. Turn off internet connection
5. Select global attribute
6. Observe error state
7. Turn on internet connection and tap "Retry"
8. Observe correctly synced options list

## Screenshots

before | after
--|--
Simulator Screen Shot - iPhone 12 - 2021-02-18 at 12 37 39|Simulator Screen Shot - iPhone 12 - 2021-02-18 at 12 38 35

Update release notes:
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Products/Variations/Add Attributes/AddAttributeOptionsViewModel.swift [hunk 1]
```diff
@@ -133,6 +147,12 @@ final class AddAttributeOptionsViewModel {
         self.stores = stores
         self.viewStorage = viewStorage
 
+        updateData()
+    }
+
+    /// Updates all data based on attribute type
+    ///
+    func updateData() {
```
Comment (Ecarrion) [inline WooCommerce/Classes/ViewRelated/Products/Variations/Add Attributes/AddAttributeOptionsViewModel.swift]: Maybe `synchronizeOptions` could be a better name?

File: WooCommerce/Classes/ViewRelated/Products/Variations/Add Attributes/AddAttributeOptionsViewModel.swift [hunk 2]
```diff
@@ -36,7 +44,7 @@ final class AddAttributeOptionsViewModel {
 
         /// Indicates if the view model is syncing a global attribute options
         ///
-        var isSyncing: Bool = false
+        var isSyncing: SyncState = .idle
```
Comment (Ecarrion) [inline WooCommerce/Classes/ViewRelated/Products/Variations/Add Attributes/AddAttributeOptionsViewModel.swift]: Probably worth it to rename this variable to `syncState` and update the documentation!

File: WooCommerce/Classes/ViewRelated/Products/Variations/Add Attributes/AddAttributeOptionsViewController.swift [hunk 3]
```diff
@@ -21,6 +21,19 @@ final class AddAttributeOptionsViewController: UIViewController {
...
+    private lazy var emptyStateViewController = EmptyStateViewController(style: .list)
+
+    /// Sync error state configuration for EmptyStateViewController
+    ///
+    private lazy var errorStateConfig: EmptyStateViewController.Config = {
+        let message = NSAttributedString(string: Localization.syncErrorMessage, attributes: [.font: EmptyStateViewController.Config.messageFont])
+        return .withButton(message: message, image: .errorImage, details: "", buttonTitle: Localization.retryAction) { [weak self] in
+            self?.viewModel.updateData()
+        }
+    }()
```
Comment (Ecarrion) [inline WooCommerce/Classes/ViewRelated/Products/Variations/Add Attributes/AddAttributeOptionsViewController.swift]: I think both of these properties can be merged into one, don't you think?

File: WooCommerce/WooCommerceTests/ViewRelated/Variations/Add Attributes/AddAttributeOptionsViewModelTests.swift [hunk 4]
```diff
@@ -207,6 +207,28 @@ final class AddAttributeOptionsViewModelTests: XCTestCase {
...
+
+        // When
+        let viewModel = AddAttributeOptionsViewModel(product: sampleProduct(), attribute: .existing(attribute: sampleAttribute()), stores: stores)
+        XCTAssertFalse(viewModel.showSyncingError)
+
+        // Then
+        waitUntil {
+            viewModel.showSyncingError
+        }
+    }
```
Comment (Ecarrion) [inline WooCommerce/WooCommerceTests/ViewRelated/Variations/Add Attributes/AddAttributeOptionsViewModelTests.swift]: ❤️

**Review by Ecarrion — COMMENTED** [review]
I really like the approach you took here with the sync state! 🚀 
Just left some comments to make the changes a bit more consistent with the rest of the project.

Also, @Garance91540 what do you think about this UX?

**Comment by ealeksandrov** [comment]
Thank you for review and great feedback, @Ecarrion! Pushed small updates.

**Review by Ecarrion — COMMENTED** [review]
Looks good to me, but let's way for @Garance91540 input on the new UX before merging!

**Review by Ecarrion — APPROVED** [review]

**Comment by Ecarrion** [comment]
> Looks good to me, but let's way for @Garance91540 input on the new UX before merging!

cc @adamzelinski

**Review by ghost — APPROVED** [review]

**Pull request merged**